### PR TITLE
opt: convert panics to assertion errors

### LINKED ISSUE #31273: opt: translate all opt panics as SQL internal errors
Currently only panics instantiated via builderError are translated to SQL errors. Every other panic will crash the server.

this is not necessary; the panics originating by explicit `panic()` calls (i.e. not implicit panics due to nil references etc) should be translated to pg errors with code "internal error", and propagated to the client.

To achieve this I foresee the following work to be necessary:

- audit the current uses of `panic()`, add the error constructors where needed. This should leverage the new `pgerror.NewAssertionError` introduced in #31261 or devise a similar mechanism which suitably populates the InternalCommand field.

- add the `recover` case that will do this translation

- add a linter, to prevent future uses of `panic()` without an error constructor

### PR DESCRIPTION
**Pull request by RaduBerinde**
This set of commits changes all components of the optimizer to throw `pgerror.NewAssertionErrorf` in panics, which are caught at the "entry points" of the optimizer and converted to errors.

I won't merge this until @knz's upcoming changes to show the stack traces for the assertion errors (and I will run some more manual tests then).

Fixes #31273, #35643.

### CODE REVIEW

**Comment by knz** [comment]
Maybe rebase this on top of #35820

**Review by andy-kimball — APPROVED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35761#-:-La6gQurEBwcAKusCe8n:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, @knz, and @rytaft)

**Comment by RaduBerinde** [comment]
I rebased locally and verified some assertion errors from inside opt, they look good! I will update the PR once that merges.

File: pkg/sql/opt/memo/statistics_builder.go [hunk 1]
```diff
@@ -346,7 +346,7 @@ func (sb *statisticsBuilder) colStat(colSet opt.ColSet, e RelExpr) *props.Column
 		return sb.colStatLeaf(colSet, &relProps.Stats, &relProps.FuncDeps, relProps.NotNullCols)
 	}
 
-	panic(fmt.Sprintf("unrecognized relational expression type: %v", e.Op()))
+	panic(pgerror.NewAssertionErrorf("unrecognized relational expression type: %v", e.Op()))
```
Comment (knz) [inline pkg/sql/opt/memo/statistics_builder.go]: Here and in other places you could have used `log.Safe()`.

Comment (RaduBerinde) [reply pkg/sql/opt/memo/statistics_builder.go]: Done, went through all of them and updated as necessary.

Comment (knz) [reply pkg/sql/opt/memo/statistics_builder.go]: Thanks! and sorry for the inconvenience.

**Comment by knz** [comment]
LGTM

**Comment by knz** [comment]
sorry there's a merge conflict you need to rebase

**Comment by danhhz** [comment]
Getting this out of the bors queue for ya

bors r-

**Comment by knz** [comment]
yay :tada: 

this will want a backport too

**Pull request merged**